substrate/frame/contracts/src/exec.rs

```diff
@@ -988,24 +993,27 @@ where
 						Event::Instantiated { deployer: caller, contract: account_id.clone() },
 					);
 				},
-				(ExportedFunction::Call, Some(code_hash)) => {
-					Contracts::<T>::deposit_event(
-						vec![T::Hashing::hash_of(account_id), T::Hashing::hash_of(&code_hash)],
-						Event::DelegateCalled { contract: account_id.clone(), code_hash },
-					);
-				},
+				(ExportedFunction::Call, Some(code_hash)) =>
+					if !self.is_read_only() {
+						Contracts::<T>::deposit_event(
+							vec![T::Hashing::hash_of(account_id), T::Hashing::hash_of(&code_hash)],
+							Event::DelegateCalled { contract: account_id.clone(), code_hash },
+						);
+					},
```

Comment:
Delegate call can't be read only. So this check is not necessary.

Reply:
I was checking it in the [evm](https://github.com/ethereum/go-ethereum/blob/d2f00cb54edc4486314c25d9e6c5b739009c2201/core/vm/instructions.go#L722)
I have not seen there such restriction. It is implemented in similar way to `opCallCode`, which is an exception in [EIP-214](https://eips.ethereum.org/EIPS/eip-214): `As an exception, CALLCODE is not considered state-changing, even with a non-zero value.`. Please let me know WDYT